// react-hichestan-datetimepicker/dist/Picker/Hours.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import React from 'react';
import { mapToFarsi } from '../dateUtils';
import { UpIcon, DownIcon } from './Icons';

var Hours =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Hours, _React$Component);

  function Hours(props) {
    var _this;

    _classCallCheck(this, Hours);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Hours).call(this, props));

    _this.start = function (e, step) {
      e.preventDefault();

      if (_this.interval) {
        clearInterval(_this.interval);
      }

      _this.fireChange(_this.props.hour + step);

      _this.interval = setInterval(function () {
        _this.fireChange(_this.props.hour + step);
      }, 200);
    };

    _this.stop = function (e) {
      e.preventDefault();

      if (_this.interval) {
        clearInterval(_this.interval);
        _this.interval = undefined;
      }
    };

    _this.fireChange = function (newHour) {
      if (newHour > 23) newHour = 0;
      if (newHour < 0) newHour = 23;

      _this.props.changeEvent(newHour);
    };

    _this.state = {};
    return _this;
  }

  _createClass(Hours, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var hour = this.props.hour;
      var hourString = hour.toString();
      hourString = mapToFarsi('00'.substring(0, 2 - hourString.length) + hourString);
      return React.createElement("div", {
        className: "JC-HourMinute"
      }, React.createElement("div", {
        className: "JC-Nav JC-Nav-Hour",
        onMouseDown: function onMouseDown(e) {
          return _this2.start(e, 1);
        },
        onMouseUp: function onMouseUp(e) {
          return _this2.stop(e);
        },
        onTouchStart: function onTouchStart(e) {
          return _this2.start(e, 1);
        },
        onTouchEnd: function onTouchEnd(e) {
          return _this2.stop(e);
        }
      }, React.createElement(UpIcon, null)), React.createElement("span", {
        className: "JC-Title",
        style: {
          textAlign: 'right'
        }
      }, hourString), React.createElement("div", {
        className: "JC-Nav JC-Nav-Hour",
        onMouseDown: function onMouseDown(e) {
          return _this2.start(e, -1);
        },
        onMouseUp: function onMouseUp(e) {
          return _this2.stop(e);
        },
        onTouchStart: function onTouchStart(e) {
          return _this2.start(e, -1);
        },
        onTouchEnd: function onTouchEnd(e) {
          return _this2.stop(e);
        }
      }, React.createElement(DownIcon, null)));
    }
  }]);

  return Hours;
}(React.Component);

export default Hours;
