// react-hichestan-datetimepicker/dist/Picker/Minutes.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import React from 'react';
import { mapToFarsi } from '../dateUtils';
import { UpIcon, DownIcon } from './Icons';

var Minutes =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Minutes, _React$Component);

  function Minutes(props) {
    var _this;

    _classCallCheck(this, Minutes);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Minutes).call(this, props));

    _this.start = function (e, step) {
      e.preventDefault();

      if (_this.interval) {
        clearInterval(_this.interval);
      }

      _this.fireChange(_this.props.minute + step);

      _this.interval = setInterval(function () {
        _this.fireChange(_this.props.minute + step);
      }, 200);
    };

    _this.stop = function (e) {
      e.preventDefault();

      if (_this.interval) {
        clearInterval(_this.interval);
        _this.interval = undefined;
      }
    };

    _this.fireChange = function (newMinute) {
      if (newMinute > 59) newMinute = 0;
      if (newMinute < 0) newMinute = 59;

      _this.props.changeEvent(newMinute);
    };

    _this.state = {};
    return _this;
  }

  _createClass(Minutes, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var minute = this.props.minute;
      var minuteString = minute.toString();
      minuteString = mapToFarsi('00'.substring(0, 2 - minuteString.length) + minuteString);
      return React.createElement("div", {
        className: "JC-HourMinute"
      }, React.createElement("div", {
        className: "JC-Nav JC-Nav-Hour",
        onMouseDown: function onMouseDown(e) {
          return _this2.start(e, 1);
        },
        onMouseUp: function onMouseUp(e) {
          return _this2.stop(e);
        },
        onTouchStart: function onTouchStart(e) {
          return _this2.start(e, 1);
        },
        onTouchEnd: function onTouchEnd(e) {
          return _this2.stop(e);
        }
      }, React.createElement(UpIcon, null)), React.createElement("span", {
        className: "JC-Title",
        style: {
          textAlign: 'left'
        }
      }, minuteString), React.createElement("div", {
        className: "JC-Nav JC-Nav-Hour",
        onMouseDown: function onMouseDown(e) {
          return _this2.start(e, -1);
        },
        onMouseUp: function onMouseUp(e) {
          return _this2.stop(e);
        },
        onTouchStart: function onTouchStart(e) {
          return _this2.start(e, -1);
        },
        onTouchEnd: function onTouchEnd(e) {
          return _this2.stop(e);
        }
      }, React.createElement(DownIcon, null)));
    }
  }]);

  return Minutes;
}(React.Component);

export default Minutes;
